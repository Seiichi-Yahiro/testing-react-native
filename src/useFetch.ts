import { useCallback, useEffect, useState } from 'react';

export enum FetchStatus {
    NotStarted = 'NotStarted',
    Pending = 'Pending',
    Successful = 'Successful',
    Failed = 'Failed',
}

export interface UseFetchData<T = any> {
    status: FetchStatus;
    retry: () => void;
    data?: T;
}
const useFetch = <T = any>(url: string): UseFetchData<T> => {
    const [status, setStatus] = useState<FetchStatus>(FetchStatus.NotStarted);
    const [data, setData] = useState<T>();

    const fetchData = useCallback(async () => {
        if (status === FetchStatus.Pending) {
            return;
        }

        setStatus(FetchStatus.Pending);

        try {
            const response = await fetch(url);

            if (response.ok) {
                const data = await response.json();
                setData(data);
                setStatus(FetchStatus.Successful);
            } else {
                setStatus(FetchStatus.Failed);
            }
        } catch (error) {
            setStatus(FetchStatus.Failed);
        }
    }, [url, status]);

    useEffect(() => {
        fetchData();
    }, [url]);

    return { status, data, retry: fetchData };
};

export default useFetch;
