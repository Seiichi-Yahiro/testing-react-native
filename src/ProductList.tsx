import React from 'react';
import {
    FlatList,
    View,
    Text,
    StyleSheet,
    Button,
    ActivityIndicator,
} from 'react-native';
import addThousandsSeparators from './addThousandsSeparators';
import useFetch, { FetchStatus } from './useFetch';

interface ProductListProps {}

const ProductList: React.FC<ProductListProps> = () => {
    const {
        data: products,
        status,
        retry,
    } = useFetch<Product[]>('https://api.example.com/products');

    const renderContent = () => {
        switch (status) {
            case FetchStatus.NotStarted:
                return;
            case FetchStatus.Pending:
                return <ActivityIndicator testID={'activityIndicator'} />;
            case FetchStatus.Successful:
                return (
                    products && (
                        <FlatList
                            data={products}
                            renderItem={Product}
                            keyExtractor={(item) => item.id}
                        />
                    )
                );
            case FetchStatus.Failed:
                return <Button title={'Retry'} onPress={retry} />;
        }
    };

    return <View style={styles.container}>{renderContent()}</View>;
};

export interface ProductProps {
    item: Product;
}
export const Product: React.FC<ProductProps> = ({ item }) => (
    <View style={styles.product}>
        <Text style={styles.productName}>{item.name}</Text>
        <Text style={styles.productPrice}>
            {addThousandsSeparators((item.price / 100).toString())}â‚¬
        </Text>
    </View>
);

export interface Product {
    id: string;
    name: string;
    price: number;
}

const styles = StyleSheet.create({
    container: {
        flex: 1,
        backgroundColor: '#fff',
        alignItems: 'center',
        justifyContent: 'center',
    },
    product: {
        borderBottomWidth: 1,
        borderBottomColor: '#ccc',
        padding: 10,
    },
    productName: {
        fontSize: 18,
        fontWeight: 'bold',
    },
    productPrice: {
        fontSize: 16,
        color: '#666',
    },
});
export default ProductList;
