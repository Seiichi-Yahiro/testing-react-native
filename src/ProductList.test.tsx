import * as useFetch from './useFetch';
import { fireEvent, render, screen } from '@testing-library/react-native';
import ProductList, { Product } from './ProductList';
import '@testing-library/jest-native/extend-expect';
import { ProductI } from './ProductListTypes';

describe('Product List', () => {
    afterEach(() => jest.restoreAllMocks());

    it('should render the data from the api', () => {
        const mockedReturnData: ProductI[] = [
            { id: '1', name: 'Apple', price: 199 },
            { id: '2', name: 'Orange', price: 299 },
        ];

        jest.spyOn(useFetch, 'default').mockReturnValueOnce({
            status: useFetch.FetchStatus.Successful,
            data: mockedReturnData,
            retry: () => {},
        });

        render(<ProductList />);

        expect(screen.getByText('Apple')).toBeOnTheScreen();
        expect(screen.getByText('Orange')).toBeOnTheScreen();

        expect(screen.getByText('1.99€')).toBeOnTheScreen();
        expect(screen.getByText('2.99€')).toBeOnTheScreen();
    });

    it('should inform when product list is empty', () => {
        jest.spyOn(useFetch, 'default').mockReturnValueOnce({
            data: [],
            status: useFetch.FetchStatus.Successful,
            retry: () => {},
        });

        render(<ProductList />);

        expect(screen.getByText('No products found')).toBeOnTheScreen();
    });

    it('should call retry on list refresh', () => {
        const retry = jest.fn();

        jest.spyOn(useFetch, 'default').mockReturnValueOnce({
            status: useFetch.FetchStatus.Failed,
            retry,
        });

        render(<ProductList />);

        fireEvent(screen.getByText('No products found'), 'refresh');

        expect(retry).toBeCalledTimes(1);
    });

    describe('Product', () => {
        it('should display the product name', () => {
            render(<Product item={{ id: '1', name: 'Foo', price: 299 }} />);

            expect(screen.getByText('Foo')).toBeOnTheScreen();
        });

        it('should display cents', () => {
            render(<Product item={{ id: '1', name: 'Foo', price: 299 }} />);

            expect(screen.getByText('2.99€')).toBeOnTheScreen();
        });

        it('should add thousand separators to the price', () => {
            render(<Product item={{ id: '1', name: 'Foo', price: 234599 }} />);

            expect(screen.getByText('2,345.99€')).toBeOnTheScreen();
        });
    });

    describe('Filter products', () => {
        it('should display all products after emptying search field', () => {
            const mockedReturnData: ProductI[] = [
                { id: '1', name: 'Apple', price: 199 },
                { id: '2', name: 'Orange', price: 299 },
                { id: '3', name: 'Apple Juice', price: 399 },
                { id: '4', name: 'Orange Juice', price: 499 },
            ];

            jest.spyOn(useFetch, 'default').mockReturnValue({
                status: useFetch.FetchStatus.Successful,
                data: mockedReturnData,
                retry: () => {},
            });

            render(<ProductList />);

            const textInput = screen.getByPlaceholderText('Search');

            fireEvent.changeText(textInput, 'Orange');
            fireEvent.changeText(textInput, '');

            expect(screen.getByText('Apple')).toBeOnTheScreen();
            expect(screen.getByText('Apple Juice')).toBeOnTheScreen();
            expect(screen.getByText('Orange')).toBeOnTheScreen();
            expect(screen.getByText('Orange Juice')).toBeOnTheScreen();
        });

        it('should search for products that contain the word Orange', () => {
            const mockedReturnData: ProductI[] = [
                { id: '1', name: 'Apple', price: 199 },
                { id: '2', name: 'Orange', price: 299 },
                { id: '3', name: 'Apple Juice', price: 399 },
                { id: '4', name: 'Orange Juice', price: 499 },
            ];

            jest.spyOn(useFetch, 'default').mockReturnValue({
                status: useFetch.FetchStatus.Successful,
                data: mockedReturnData,
                retry: () => {},
            });

            render(<ProductList />);

            const apple = screen.getByText('Apple');
            const appleJuice = screen.getByText('Apple Juice');

            const orange = screen.getByText('Orange');
            const orangeJuice = screen.getByText('Orange Juice');

            fireEvent.changeText(
                screen.getByPlaceholderText('Search'),
                'Orange'
            );

            expect(orange).toBeOnTheScreen();
            expect(orangeJuice).toBeOnTheScreen();
            expect(apple).not.toBeOnTheScreen();
            expect(appleJuice).not.toBeOnTheScreen();
        });
    });
});
