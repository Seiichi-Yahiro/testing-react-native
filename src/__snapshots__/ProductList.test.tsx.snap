// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product List Snapshot Product snapshot 1`] = `
<View
  style={
    {
      "borderBottomColor": "#ccc",
      "borderBottomWidth": 1,
      "padding": 10,
    }
  }
>
  <Text
    style={
      {
        "fontSize": 18,
        "fontWeight": "bold",
      }
    }
  >
    Foo
  </Text>
  <Text
    style={
      {
        "color": "#666",
        "fontSize": 16,
      }
    }
  >
    2,345.99
    â‚¬
  </Text>
</View>
`;
